amazon/ion/ioncmodule.c

```diff
@@ -792,6 +794,28 @@ iERR ionc_write_value(hWRITER writer, PyObject* obj, PyObject* tuple_as_sexp) {
         IONCHECK(ionc_write_sequence(writer, obj, tuple_as_sexp));
         IONCHECK(ion_writer_finish_container(writer));
     }
+    /* When we are going to serialize a lazy Ion object, we should check if this object holds a cached binary
+    representation already. If so, write the cached bytes into writer's temporary buffer directly. However I didn't find
+    an API in ion-c's ion_writer.h to append arbitrary bytes in writer's temporary buffer (correct me if I missed it) so
+    I write an IonBlob value instead, which is wrong! It's only used for visible and debugging. Later on we should decide
+    if/how to expose the existing API in ion-c for CPython layer so that it's able to append custom bytes into writer's
+    value_stream instead of only be able to call the writer to writes bytes.
+     */
+    else if (PyObject_IsInstance(obj, _ionpylazyobj_cls)) {
+        Py_ssize_t len;
+        char* c_buf;
+
+        // TODO check the existence of lazy buffer first. Since all lazy objects hold a buffer by my sample code... I skipped validation
+        PyObject* lazy_buffer = PyObject_GetAttrString(obj, "lazy_buffer");
+
+        // Convert python bytes to C bytes
+        if (PyBytes_AsStringAndSize(lazy_buffer, &c_buf, &len) < 0) {
+            _FAILWITHMSG(IERR_INVALID_ARG, "Binary conversion error.");
+        }
+        // TODO should append the c_buf to _value_stream directly. Not an Ion BLOB!
+        // E.g. Something like ION_PUT(writer->_typed_writer.binary._value_stream, c_buf);
+        IONCHECK(ion_writer_write_blob(writer, (BYTE*)c_buf, len));
```

Comment:
Thanks, it works. But I need to investigate why ion_writer_flush() didn't work in my sample code so the output is not in a right order... 

Current logic parses a list Lazy object to an Ion list including child lazy objects, we don't know the total length when we append the cached bytes in the underlying ion_stream so it requires a little more calculation. 

To know how to calculate the length, the question then turns to how we handle a lazy list. My idea is:
1. Simply create a lazy object, cache the list's binary representation and set the type to `ION_TYPE.LIST` like other Ion types. Note that all lazy objects are immutable.
2. When users only need to append/delete objects, they can `partially_deserialize` the Lazy object into a list holding all child lazy objects in order. They can add new Ion values or delete existing values.
3. When users need to modify a list and its child Ion values, they can `fully_deserialize` the lazy object into a regular Ion list class holding all deserialized Ion values. They can add, delete or modify values.

The second point above partially deserialize the list so we can reuse the cached binary of its child lazy objects. In other words, user just need to deserialize desired Ion values instead of a full list. Later on if users want to modify a specific child lazy object within the List/Sexp, they can deserialize that lazy object, modify it, and replaced the original lazy object with the modified Ion value. This still requires a little more experiment, if partial deserialization and the strategy of modifying lazy objects is doable and faster, there is no need to do a fully deserialization. In addition we should consider Ion struct situation since it is unordered and allows duplicated keys.

Reply:
> But I need to investigate why ion_writer_flush() didn't work in my sample code so the output is not in a right order

Solved, I forgot to call it in the top-level.